File: leetcode/0050 Pow(x,n).py
"""
* 50. Pow(x, n) （快速幂，清晰图解） - Pow(x, n) - 力扣（LeetCode） 
    * https://leetcode-cn.com/problems/powx-n/solution/50-powx-n-kuai-su-mi-qing-xi-tu-jie-by-jyd/
"""

class Solution:
    def myPow(self, x: float, n: int) -> float:
        if x==0.0:
            return 0.0
        if n<0:
            x,n = 1/x, -n

        res = 1
        while n:
            if n&1:
                res *= x
            n >>=1
            if n:
                x *= x
        return res

File: 面试高频题/0050 pow(x,n).cpp
class Solution {
public:
    double myPow(double x, int n) {
        if (x==0)
            return 0;
        double res=1;
        if (n<0) {
            if (n==INT_MIN) { // 最小的负数会溢出
                return myPow(x,n+1) * x;
            } else {
                x = 1/x;
                n = -n;
            }
        }
        while (n) {
            if (n&1) { // 如果该位为 1，则将 res 乘上此时的 x
                res *= x;
            }
            x *= x;
            n >>= 1;
        }
        return res;
    }
};